src/core/image_resizer.js

```diff
@@ -0,0 +1,389 @@
+/* Copyright 2023 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FeatureTest, ImageKind, shadow } from "../shared/util.js";
+
+const MIN_IMAGE_DIM = 2048;
+
+// In Chrome, there aren't max dimensions but only a max area. So an image with
+// a very large dimensions is acceptable but it probably doesn't hurt to reduce
+// it when considering that it will finally rendered on a small canvas.
+const MAX_IMAGE_DIM = 65537;
+const MAX_ERROR = 128;
+
+// Large images are encoded in using the BMP format (it's a way faster than
+// encoding in PNG because there are no checksums to compute).
+// Unfortunately, createImageBitmap will create a task in the main thread in
+// order to build the bitmap, hence the main thread is blocked during the
+// decoding which can be a bit long (see bug 1817244).
+// When bug 1759728 is fixed we can remove the BMP workaround and at least it
+// should be a way faster to create the bitmap.
+
+class ImageResizer {
+  constructor(imgData, isMask) {
+    this._imgData = imgData;
+    this._isMask = isMask;
+  }
+
+  static needsToBeResized(width, height) {
+    if (width <= this._goodSquareLength && height <= this._goodSquareLength) {
+      return false;
+    }
+
+    const { MAX_DIM } = this;
+    if (width > MAX_DIM || height > MAX_DIM) {
+      return true;
+    }
+
+    const area = width * height;
+    if (this._hasMaxArea) {
+      return area > this.MAX_AREA;
+    }
+
+    if (area < this._goodSquareLength ** 2) {
+      return false;
+    }
+
+    // We try as much as possible to avoid to compute the max area.
+    if (this._areGoodDims(width, height)) {
+      this._goodSquareLength = Math.max(
+        this._goodSquareLength,
+        Math.floor(Math.sqrt(width * height))
+      );
+      return false;
+    }
+
+    // TODO: the computation can be a bit long because we potentially allocate
+    // some large canvas, so in the Firefox case this value (and MAX_DIM) can be
+    // infered from prefs (MAX_AREA = gfx.max-alloc-size / 4, 4 is because of
+    // RGBA).
+    this._goodSquareLength = this._guessMax(
+      this._goodSquareLength,
+      MAX_DIM,
+      MAX_ERROR,
+      0
+    );
+    this._hasMaxArea = true;
+    const maxArea = this._goodSquareLength ** 2;
+
+    shadow(this, "MAX_AREA", maxArea);
+
+    return area > maxArea;
```

Comment:
Can we please utilize the setter instead, to reduce the amount of manual handling here?
```suggestion
    this.MAX_AREA = this._goodSquareLength ** 2;
    
    return area > this.MAX_AREA;
```